specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+
+## Registration of the Push Activator for Win32 (Inproc)
+If a Win32 app needs to subscribe to Push triggers inproc, the code in Main would follow the pattern below:
+* The Registration will take in a CLSID of the COM server as part of the Activator setup. The registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and abstract away the COM details from the developer.
+* The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a seperate AppLifeCycle API spec.)
+* If the Activation is Push, the app will QI the Arguments property to retrieve an instance of PushActivatedEventArgs and get the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+  * It will subscribe to a Foreground In-memory Push event
+  * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler.
+    * Expect Channel operations to take around 2-16 minutes in some rare cases (retryable errors). In the 90th percentile case, it should be fairly quick operation (in a few seconds).  
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    winrt::guid activatorGuid{ 0xC380465D, 0x2271, 0x428C, { 0x9B, 0x83, 0xEC, 0xEA, 0x3B, 0x4A, 0x85, 0xC1} };
+    InProcActivatorDetails details(activatorGuid);
+    PushManager::RegisterActivator(details);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    // Note: This API is currently not present in Reunion but will be included as part of the AppLifecycle Public API spec.
+    auto activation = AppLifecycle::Activation();
+    auto kind = activation.Kind();
+
+    if (kind == ReunionActivationKind::Push)
+    {
+        auto pushArgs = activation.Arguments().try_as<PushActivatedEventArgs>();
+
+        // Call TakeDeferral to ensure that code runs in low power
+        pushArgs.GetDeferral();
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ReunionActivationKind::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushActivated([](const auto&, PushActivatedEventArgs args)
+        {
+                // Call TakeDeferral to ensure that code runs in low power
+                args.GetDeferral();
+
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+
+                // Call CompleteDeferral as good practise: Needed mainly for low power usage
+                args.CompleteDeferral();
```

Comment:
In the current Windows SDK, with this in-memory event handler, we provided a way for developers to choose to cancel the background task activation if they want to intercept via the in-memory handler. Should we have a way to do the same here too?

**Old platform code**

```csharp
channelOld.PushNotificationReceived += (PushNotificationChannel sender, PushNotificationReceivedEventArgs args) =>
{
    if (args.RawNotification.Content.Contains("tacos"))
    {
        // Want to handle this in my currently running app
        args.Cancel = true;
    }
    else
    {
        // Otherwise, I'll let my background task handle it
    }
};
```

Reply:
I'm actually thinking that we use the Background Task Pipe itself instead of the in-memory event handler hanging off channels for the PushReceived event. This greatly simplifies the code, in the sense that developers don't have to handle the Cancellation case.